pkg/graveler/committed/prefix_iterator.go

```diff
@@ -0,0 +1,166 @@
+	"github.com/treeverse/lakefs/pkg/graveler"
+)
+
+type rangeValue struct {
+	r  *Range
+	vr *graveler.ValueRecord
+}
+
+type ImportIterator interface {
+	IsCurrentRangeBoundedByPath() bool
+	IsCurrentPathIncludedInRange() bool
+}
+
+type PrefixIterator struct {
+	currentPrefixIndex int
+	prefixes           []graveler.Prefix
+	rangeIterator      Iterator
+	currentRangeValue  rangeValue
+}
+
+func (ipi *PrefixIterator) Value() (*graveler.ValueRecord, *Range) {
+	return ipi.currentRangeValue.vr, ipi.currentRangeValue.r
+}
+
+func (ipi *PrefixIterator) updateValue() (*graveler.ValueRecord, *Range) {
+	vr, r := ipi.rangeIterator.Value()
+	ipi.currentRangeValue = rangeValue{
+		r,
+		vr,
+	}
+	return vr, r
+}
+
+func (ipi *PrefixIterator) Err() error {
+	return ipi.rangeIterator.Err()
+}
+func (ipi *PrefixIterator) Close() {
+	ipi.rangeIterator.Close()
+}
+func (ipi *PrefixIterator) SeekGE(id graveler.Key) {
+	ipi.rangeIterator.SeekGE(id)
+}
+
+func (ipi *PrefixIterator) Next() bool {
+	hasNext := ipi.rangeIterator.Next()
+	if !hasNext {
+		return false
+	}
+	vr, r := ipi.updateValue()
+	ipi.updatePath()
+
+	if vr == nil && r != nil { // head of range
+		for ipi.IsCurrentRangeBoundedByPath() {
+			if ipi.rangeIterator.Err() != nil {
+				return false
+			}
+			hasNext = ipi.rangeIterator.NextRange()
+			if !hasNext {
+				return false
+			}
+			ipi.updateValue()
+			ipi.updatePath()
+		}
+	} else {
+		prefixLen := len(ipi.prefixes)
+		for vr != nil && ipi.currentPrefixIndex < prefixLen && strings.HasPrefix(vr.Key.String(), string(ipi.prefixes[ipi.currentPrefixIndex])) {
+			if ipi.rangeIterator.Err() != nil {
+				return false
+			}
+			hasNext = ipi.rangeIterator.Next()
+			if !hasNext {
+				return false
+			}
+			vr, _ = ipi.updateValue()
+			ipi.updatePath()
+		}
+	}
+	return true
+}
+
+func (ipi *PrefixIterator) NextRange() bool {
+	hasNext := ipi.rangeIterator.NextRange()
+	if !hasNext {
+		return false
+	}
+	ipi.updateValue()
+	ipi.updatePath()
+
+	for ipi.IsCurrentRangeBoundedByPath() {
+		if ipi.rangeIterator.Err() != nil {
+			return false
+		}
+		hasNext = ipi.rangeIterator.NextRange()
+		if !hasNext {
+			return false
+		}
+		ipi.updateValue()
+		ipi.currentPrefixIndex++
+	}
+	return true
+}
+
+func NewPrefixIterator(prefixes []graveler.Prefix, rangeIterator Iterator) *PrefixIterator {
+	sort.Slice(prefixes, func(i, j int) bool {
+		return prefixes[i] < prefixes[j]
+	})
+	return &PrefixIterator{prefixes: prefixes, currentPrefixIndex: 0, rangeIterator: rangeIterator}
+}
+
+func (ipi *PrefixIterator) updatePath() {
+	if ipi.currentPrefixIndex >= len(ipi.prefixes) {
+		return
+	}
+	currMinKey := string(ipi.currentRangeValue.r.MinKey)
+	if ipi.currentRangeValue.vr != nil {
+		currMinKey = string(ipi.currentRangeValue.vr.Key)
+	}
+	// If the position is smaller or doesn't have the prefix of the current key, get the next prefix.
+	// At the end of this function we'll have either a path that is a prefix of the current key, or bigger than the key
```

Comment:
Can you reword this into a comment for what this function does?  It is hard to understand its action purely from the name.